torch/utils/data/_utils/collate.py

```diff
@@ -82,7 +85,110 @@ def default_convert(data):
+            return {key: collate([d[key] for d in batch], collate_fn_map=collate_fn_map) for key in elem}
+    elif isinstance(elem, tuple) and hasattr(elem, '_fields'):  # namedtuple
+        return elem_type(*(collate(samples, collate_fn_map=collate_fn_map) for samples in zip(*batch)))
+    elif isinstance(elem, collections.abc.Sequence):
+        # check to make sure that the elements in batch have consistent size
+        it = iter(batch)
+        elem_size = len(next(it))
+        if not all(len(elem) == elem_size for elem in it):
+            raise RuntimeError('each element in list of batch should be of equal size')
+        transposed = list(zip(*batch))  # It may be accessed twice, so we use a list.
+
+        if isinstance(elem, tuple):
+            return [collate(samples, collate_fn_map=collate_fn_map) for samples in transposed]  # Backwards compatibility.
+        else:
+            try:
+                return elem_type([collate(samples, collate_fn_map=collate_fn_map) for samples in transposed])
+            except TypeError:
+                # The sequence type may not support `__init__(iterable)` (e.g., `range`).
+                return [collate(samples, collate_fn_map=collate_fn_map) for samples in transposed]
+
+    raise TypeError(default_collate_err_msg_format.format(elem_type))
+
+
+def _collate_tensor_fn(batch, *, collate_fn_map: Optional[Dict[Union[Type, Tuple[Type, ...]], Callable]] = None):
+    elem = batch[0]
+    out = None
+    if torch.utils.data.get_worker_info() is not None:
+        # If we're in a background process, concatenate directly into a
+        # shared memory tensor to avoid an extra copy
+        numel = sum(x.numel() for x in batch)
+        storage = elem.storage()._new_shared(numel, device=elem.device)
+        out = elem.new(storage).resize_(len(batch), *list(elem.size()))
+    return torch.stack(batch, 0, out=out)
+
+
+def _collate_numpy_array_fn(batch, *, collate_fn_map: Optional[Dict[Union[Type, Tuple[Type, ...]], Callable]] = None):
+    elem = batch[0]
+    # array of string classes and object
+    if np_str_obj_array_pattern.search(elem.dtype.str) is not None:
+        raise TypeError(default_collate_err_msg_format.format(elem.dtype))
+
+    return collate([torch.as_tensor(b) for b in batch], collate_fn_map=collate_fn_map)
+
+
+def _collate_numpy_scalar_fn(batch, *, collate_fn_map: Optional[Dict[Union[Type, Tuple[Type, ...]], Callable]] = None):
+    return torch.as_tensor(batch)
+
+
+def _collate_float_fn(batch, *, collate_fn_map: Optional[Dict[Union[Type, Tuple[Type, ...]], Callable]] = None):
+    return torch.tensor(batch, dtype=torch.float64)
+
+
+def _collate_int_fn(batch, *, collate_fn_map: Optional[Dict[Union[Type, Tuple[Type, ...]], Callable]] = None):
+    return torch.tensor(batch)
+
+
+def _collate_str_fn(batch, *, collate_fn_map: Optional[Dict[Union[Type, Tuple[Type, ...]], Callable]] = None):
+    return batch
+
+
+_default_collate_fn_map = {torch.Tensor: _collate_tensor_fn}
```

Comment:
We (`torchvision`) either need that public or the `_collate_*_fn`'s since we can't extend otherwise. By public I mean not necessary publicly exposed, but with guaranteed BC in future releases. This is generally not the case for private stuff and we can't depend on it.

Reply:
You probably need to annotate like this

```suggestion
_default_collate_fn_map: Dict[Type, Callable] = {torch.Tensor: _collate_tensor_fn}
```

Otherwise `mypy` makes the types more concrete and this leads to errors.

Reply:
I can remove all underscores as they are all under a private module `_utils`. And, for torchvision use case, do you need to re-order the `_default_collate_fn_map`? For example, I would expect you want to add a registry collate function before `collate_tensor_fn`. If so, you might need to create your own `collate_fn_map` rather than extending `_default_collate_fn_map`

Reply:
> And, for torchvision use case, do you need to re-order the `_default_collate_fn_map`?

Correct. But whether you let us import the map or the individual functions is somewhat irrelevant. The main issue I have here is BC. If we depend on these internal functions, we need a guarantee that BC will be kept although they are marked as "private".

Reply:
Sure. As far as I am concerned, we will support BC for those functions.